.github/workflows/pr-automation.yml

```
@@ -125,10 +133,25 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        shard: [1/4, 2/4, 3/4, 4/4]
+        shard: ${{ steps.generate-shard-matrix.outputs.matrix }}
+    with:
+      concurrency: ${{ inputs.shards || 4 }}
     steps:
       - uses: actions/checkout@v4
 
+      - name: Generate shard matrix 
+        id: generate-shard-matrix
+        run: |
+          max_concurrency=4
+          if [ "$concurrency" -gt "$max_concurrency" ]; then
+            echo "[WARN] The requested number of shards ($concurrency) exceeds maximum allowed value ($max_concurrency). Using $max_concurrency shards instead."
```

Should go to stderr instead of stdout but it's really a major issue:
```suggestion
            echo "[WARN] The requested number of shards ($concurrency) exceeds maximum allowed value ($max_concurrency). Using $max_concurrency shards instead." >&2
```